src/dune_rules/melange/melange_rules.ml

```diff
@@ -52,6 +52,21 @@ let impl_only_modules_defined_in_this_lib sctx lib =
           (Lib.name lib |> Lib_name.to_string)
       ]
   | Some modules ->
+    let () =
+      let info = Lib.info lib in
+      let modes = Lib_info.modes info in
+      match modes.melange with
+      | false ->
+        User_error.raise
+          ~loc:(fst mel.libraries)
```

Comment:
It would be nice to show the specific location for the library in question under the `libraries` field, but I understand that:

- the library may not be present because it's transitively depended upon
- we may not have that original loc at this point?

Reply:
Great point. This comment made me think we should also test the case for failures where the incompatible library is depended upon transitively. The results were very confusing, see test updates in https://github.com/ocaml/dune/pull/6412/commits/bb3fccbf9fbd3de8005d9eb506c2f71ecf7ca4df.

So I ended up implementing your suggestion, all the information about the incompatible lib location was readily available, and the diff becomes way simpler now, see https://github.com/ocaml/dune/pull/6412/commits/486a64359962cc5da63990e34507b0719165973c. Lmk what you think!

Reply:
> So I ended up implementing your suggestion

After another read, I am not sure this was the suggestion 😅 but I think it's better to point users to the faulty library location, rather than some initial or intermediate part of the dependency chain, which could be more confusing. In a future improvement, if we end up realizing this is confusing, we could try to print the whole chain in the error message, somehow.

Reply:
That does indeed point to a more specific place. And the diff is indeed way simpler. This is fine for me.